pkg/protocol/block_dispatcher.go

```diff
@@ -205,21 +214,130 @@ func (b *BlockDispatcher) processWarpSyncResponse(commitmentID iotago.Commitment
+	// As we already decided to switch and sync to this chain we should make sure that processing the blocks from the commitment
+	// leads to the verified commitment.
+	if targetEngine.Notarization.AcceptedBlocksCount(commitmentID.Slot()) > 0 {
+		b.protocol.activeEngineMutex.Lock()
+
+		newEngine, err := b.protocol.EngineManager.RollbackEngine(commitmentID.Slot() - 1)
+		if err != nil {
+			return ierrors.Wrapf(err, "failed to rollback engine for slot %d", commitmentID.Slot())
+		}
+		b.protocol.Events.MainEngineRestarted.Trigger(newEngine)
+
+		newEngine.SetChainID(targetEngine.ChainID())
+
+		if err := b.protocol.EngineManager.SetActiveInstance(newEngine); err != nil {
+			return ierrors.Wrap(err, "failed to set active engine instance")
+		}
+		//
+		b.protocol.linkToEngine(newEngine)
+		//
+		b.protocol.mainEngine.Shutdown()
+		b.protocol.mainEngine = newEngine
+		targetEngine = newEngine
+		b.protocol.activeEngineMutex.Unlock()
+	}
+
+	// Once all blocks are booked we
+	//   1. Mark all transactions as accepted
+	//   2. Mark all blocks as accepted
+	//   3. Force commitment of the slot
+	totalBlocks := uint32(len(blockIDs))
+	var bookedBlocks atomic.Uint32
+	var notarizedBlocks atomic.Uint32
+
+	forceCommitmentFunc := func() {
+		// 3. Force commitment of the slot
+		producedCommitment, err := targetEngine.Notarization.ForceCommit(commitmentID.Slot())
+		if err != nil {
+			b.protocol.HandleError(err)
+			return
+		}
+
+		// 4. Verify that the produced commitment is the same as the initially requested one
+		if producedCommitment.ID() != commitmentID {
+			b.protocol.HandleError(ierrors.Errorf("producedCommitment ID mismatch: %s != %s", producedCommitment.ID(), commitmentID))
+			return
+		}
+	}
+
+	blockBookedFunc := func(_, _ bool) {
+		if bookedBlocks.Add(1) != totalBlocks {
+			return
+		}
+
+		// 1. Mark all transactions as accepted
+		for _, transactionID := range transactionIDs {
+			targetEngine.Ledger.ConflictDAG().SetAccepted(transactionID)
+		}
+
+		// 2. Mark all blocks as accepted
+		for _, blockID := range blockIDs {
+			block, exists := targetEngine.BlockCache.Block(blockID)
+			if !exists { // this should never happen as we just booked these blocks in this slot.
+				continue
+			}
+
+			targetEngine.BlockGadget.SetAccepted(block)
+
+			block.Notarized().OnUpdate(func(_, _ bool) {
+				// Wait for all blocks to be notarized before forcing the commitment of the slot.
+				if notarizedBlocks.Add(1) != totalBlocks {
+					return
+				}
+
+				forceCommitmentFunc()
+			})
+		}
+	}
+
+	if len(blockIDs) == 0 {
+		forceCommitmentFunc()
+		return nil
+	}
+
 	for _, blockID := range blockIDs {
-		targetEngine.BlockDAG.GetOrRequestBlock(blockID)
+		block, _ := targetEngine.BlockDAG.GetOrRequestBlock(blockID)
+		if block == nil { // this should never happen as we're requesting the blocks for this slot so it can't be evicted.
+			continue
```

Comment:
Should we panic or error then? This can only legitimally be the case if there no blocks for this slot, but you already capture this condition above.

Reply:
If there's no blocks for the slot then the loop would never be entered. Panicing is bad, logging and skipping seems the best imo. I added an error handler which should ultimatley be replaced with proper logging (this could be a debug).